### PR TITLE
* Improve init script: specifically manage salt configurations rather than arbitrary salt processes

### PR DESCRIPTION
**Pull request by plastikos**
### What does this PR do?

Prevents SysV init script from killing non-salt processes.
### What issues does this PR fix or reference?

None
### Previous Behavior

Current SysV init script kills unrelated, non-salt processes.
### New Behavior

Manage a specific salt minion configuration rather than anything that seems to be a salt minion.
### Tests written?

Yes

Unfortunately SysV init scripts tend to rummage through PIDs filtering for
appropriate processes to manage.  Unfortunately the filters are usually weak
and don't account for similar processes run by other users, PIDs of dead
processes being re-used for completely different executables, etc..  These
weaknesses can result in killing unrelated processes with potentially serious
results.

These improvements to the SysV init script is a complete rewrite with the
following improvements:
- Specifically manage individual salt configurations rather than looking for
  salt minion-like processes.
- Obtain salt minion information from the salt configuration - use the
  information to manage the specifically configured process.
- Drop all of the platform-specific helper functions that allow the
  previously-mentioned weaknesses.
  - Unfortunately this means that the output information may not match the
    specific platform (this could easily be corrected).
- Now can manage multiple salt processes started by different users
  - Unfortunately starts/stops/restarts as a group and is unable to manage
    them both as a group or as individual processes (this could easily be
    corrected)

The new initscript also allows various control variables to be overridden by
environment variables or through settings put in `/etc/sysconf/salt` or
`/etc/default/salt`.

:SALTMINION_DEBUG: Dump each line expansion before execution, output system
                   information on failure.  Default: unset

:SALTMINION_BINDIR: Location of `salt-minion`, `salt-call` and other
                    executables.  Default: `/usr/bin`

:SALTMINION_SYSCONFDIR: The parent directory for the `salt` configuration
                        directory and the `sysconfig` or `default`
                        directory.

Rebased P.R. #32025

### CODE REVIEW

**Comment by cachedout** [comment]
This looks really good to me. +1

I'd also like @terminalmage and @dmurphy18 to take a look. Probably @s0undt3ch as well.

**Comment by plastikos** [comment]
There are a few things that are sub-optimal:

* The ``salt-call`` invocation to query settings from the config is slow.
* If there are multiple, simultaneous invocations of ``salt-minion`` there is a race between the invocation and the appearance of the unix-domain sockets.  That could be mitigated with a judicious file lock.

At this point I'm content that this is better than before without being too over-engineered.  Potentially SysV init scripts are on the decline with the proliferation of ``systemd``, but it still merits having this improved.  This was initiated because the previous script was klling ``mysqld`` - which ultimately is a problem with the RedHat helper functions and not with ``salt-minion`` in particular.  Changing ``salt-minion`` is more fixable than the helper functions.

**Comment by plastikos** [comment]
Bump.

**Comment by vutny** [comment]
I see that `SALTMINION_BINDIR` and `SALTMINION_SYSCONFDIR` variables are being used before they were defined or sourced from elsewhere... And I can't find where default values have been set.

File: pkg/rpm/salt-minion [hunk 1]
```diff
@@ -25,132 +25,296 @@
...
-SALTMINION=/usr/bin/salt-minion
-PYTHON=/usr/bin/python
-MINION_ARGS=""
+SERVICE="salt-minion"
+PROCESS="salt-minion"
 
-if [ -f /etc/default/salt ]; then
-    . /etc/default/salt
+# Read in settings file
+if [ -f "${SALTMINION_SYSCONFDIR}/default/salt" ]; then
```
Comment (vutny) [inline pkg/rpm/salt-minion]: `salt-minion` RPM already comes with `/etc/default/salt-minion` file by the way. Why not use it?

**Comment by plastikos** [comment]
@vutny , I'm a bit dumbfounded.  I know those were set, as well as ``SALTMINION_PYTHON``, in a previous version - I have been using it in production.  I must have done something really dumb when I rebased and squashed.  I'll check it in the morning.

Thanks for spotting such a bone-headed move.

**Comment by plastikos** [comment]
@vutney, fixed the missing lines.

**Comment by plastikos** [comment]
@cachedout, thanks!

**Pull request merged**